### PR TITLE
[SPARK-29003][CORE] Add `start` method to ApplicationHistoryProvider to avoid deadlock on startup

### PR DESCRIPTION
**Pull request by shanyu**
### What changes were proposed in this pull request?

During Spark History Server startup, there are two things happening simultaneously that call into `java.nio.file.FileSystems.getDefault()` and we sometime hit [JDK-8194653](https://bugs.openjdk.java.net/browse/JDK-8194653).
1) start jetty server
2) start ApplicationHistoryProvider (which reads files from HDFS)

We should do these two things sequentially instead of in parallel.
We introduce a start() method in ApplicationHistoryProvider (and its subclass FsHistoryProvider), and we do initialize inside the start() method instead of the constructor.
In HistoryServer, we explicitly call provider.start() after we call bind() which starts the Jetty server.

### Why are the changes needed?
It is a bug that occasionally starting Spark History Server results in process hang due to deadlock among threads.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
I stress tested this PR with a bash script to stop and start Spark History Server more than 1000 times, it worked fine. Previously I can only do the stop/start loop less than 10 times before I hit the deadlock issue.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 1]
```diff
@@ -200,7 +200,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
-  val initThread = initialize()
+  var initThread:Thread = null
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: nit: space between `:` and `Thread`.

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 2]
```diff
@@ -151,6 +151,7 @@ class HistoryServer(
   /** Bind to the HTTP server behind this web interface. */
   override def bind() {
     super.bind()
+    provider.start()
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: I'd rather place `provider.start()` after `server.bind()` in `main()`, not here. It feels really odd to start provider when calling bind.

**Review by HeartSaVioR — COMMENTED** [review]
The approach looks OK to me, though I'd call provider.start() to different place - commented.

Left some comments. There's some simpler workaround (like pre-calling `FileSystem.getDefault()` in main method of HistoryServer) so reviewer can weigh which approach makes more sense.

Before that, could you change the title of PR to respect the format, like below?
```
[SPARK-29003][CORE] Fix deadlock in startup of Spark History Server
```

Most of titles in Spark PRs are explaining what the patch is addressing, not the symptom of the issue.

**Comment by felixcheung** [comment]
what he says ^^

**Comment by HeartSaVioR** [comment]
As I commented in JIRA issue, it seems to be a known issue in JDK 8, though the fixed version is too high (8u231) which Spark cannot always expect on this. So good to have workaround.

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Hi, @shanyu .
There is a Scala style issue. Could you run `dev/scalastyle` in your PR directory and fix?
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala:203:16: No space after token :
```

**Comment by shanyu** [comment]
Will do. Thanks.

Sorry for the delay, for some reason, I don't get email update when my PR is reviewed and comments left here...

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala (hunk 2 above)
Comment (shanyu) [reply core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: That makes sense. Done.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 1 above)
Comment (shanyu) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Done.

**Comment by shanyu** [comment]
Addressed all feedback. Regarding to the point of simpler workaround raised by @HeartSaVioR, I feel that adding start() to provider seems more appropriate. There should be a way to control when to start provider as it is odd today that there is a stop() but not a start() method for provider. And calling FileSystems.getDefault() in main function seems strange and hacky.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -200,7 +200,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
-  val initThread = initialize()
+  var initThread : Thread = null
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: sorry nit again: no space between `d` and `:`. Please refer `replayExecutor` for how to space.

**Review by HeartSaVioR — APPROVED** [review]
LGTM except nit.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (shanyu) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Done.

**Review by HeartSaVioR — CHANGES_REQUESTED** [review]
Ah I forgot the test suites - test failures look relevant.

There's FsHistoryProvider in HistoryServerSuite and we haven't modify the test to call `start()`. 
Could you fix the case, as well as search all cases initializing ApplicationHistoryProvider and its descendant to call `start()` properly?

**Comment by HeartSaVioR** [comment]
Hmm... you may need to search and fix manually instead of just fixing the case in failed suite. (Or maybe you may need to run tests in your dev. environment at least for `core` module.) 

I can easily find more spots to fix. Mostly you may need to call `start()` wherever instances of descendants of ApplicationHistoryProvider class are initialized.

**Comment by shanyu** [comment]
I searched the code base and I didn't see other descendants of ApplicationHistoryProvider. The latest build passed all tests in "org.apache.spark.deploy.history", but it says "fails due to an unknown error code, -9", I have no idea what this means. Could it be an infra issue?

**Comment by HeartSaVioR** [comment]
You can find all implementations for ApplicationHistoryProvider in IDE, for me, IntelliJ. Other than FsHistoryProvider, they're in test suites.

Seems like you missed to add start() for all places initializing FsHistoryProvider, which means calling startPolling() is missing: for testing, background threads will not be launched so that's OK, but initializing disk manager and validating log directory are missing. Test suites don't concern about it so test passes, but given it's unintentional to not add start(), we would be better to fix that.

**Comment by HeartSaVioR** [comment]
Just submitted a PR https://github.com/shanyu/spark/pull/2 which addresses missing spots.

**Comment by HeartSaVioR** [comment]
>  it says "fails due to an unknown error code, -9", I have no idea what this means. Could it be an infra issue?

That might be intermittent failure and we can run the build again. If you merge my PR and push to the branch the build will rerun, so no worries.

**Comment by shanyu** [comment]
Thank you @HeartSaVioR!
I have added the provider.start() to HistoryServerSuite, it is in my PR.
I didn't add them to FsHistoryProvider because I see no need to, as you mentioned already. The test suite knows the implementation so it does not have to follow the start/stop pattern. For one thing the tests didn't call stop(). If we are to modify the implementation of FsHistoryProvider.start() we can add start() at that point.

**Comment by HeartSaVioR** [comment]
Hmm... yes I agree it might be weird if we don't follow start/stop pattern. What I'm concerned about is the change would also make tests skip initializing. If the test calls `updateAndCheck()`, it misses initialization. It doesn't let test be broken, but brings some unintentional side-effect.

I think we even need to call `initialize` instead if it's weird for us to call `start` without calling `stop`, but given tests are passing, let's hear more voice on committers.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by shanyu** [comment]
> What I'm concerned about is the change would also make tests skip initializing. If the test calls `updateAndCheck()`, it misses initialization. It doesn't let existing tests be broken, but brings some unintentional side-effect.
Calling updateAndCheck() without start() in FsHistoryProviderSuite is fine because start() just waits for safe mode exit and for this suite of tests it is irrelevant.

**Comment by HeartSaVioR** [comment]
> start() just waits for safe mode exit and for this suite of tests it is irrelevant.

No. There're some initializations happening in `startPolling()`, and it's only called once we call `initialize()`. 

https://github.com/apache/spark/blob/7f36cd2aa5e066a807d498b8c51645b136f08a75/core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala#L243-L284

Safe mode exit is just one of functionalities in `initialize()`. So while we might be able to say it doesn't break the existing tests so it's kind of safe, it still brings side-effect.

**Comment by shanyu** [comment]
@HeartSaVioR Yeah you are right, it also calls startPolling() after exit from safe mode. That's why we need it in HistoryServerSuite tests. For FsHistoryProviderSuite, we always force update so it should be OK.

**Comment by HeartSaVioR** [comment]
I'm thinking out logically, we're just missing to initialize FsHistoryProvider and its descendants for some of tests, regardless of the fact the initialization doesn't break actual tests. (`diskManager.initialize` is not a "must" thing to call so tests are passing, but what if it's a "must"?) 

It doesn't seem to be logically correct to not initialize just because tests are passing, but that's only me, and remaining code change itself is good so it's not a blocker (maybe minor?) and I'll leave a decision to committers. Let's wait and hear feedbacks.

**Comment by HeartSaVioR** [comment]
cc. @felixcheung @dongjoon-hyun Could you revisit this PR? Thanks in advance!

File: core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala [hunk 4]
```diff
@@ -114,6 +114,8 @@ private[history] abstract class ApplicationHistoryProvider {
    */
   def stop(): Unit = { }
 
+  def start(): Unit = { }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala]: nit. Could you add a function description like `def stop()`?

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM with one nit comment. This is fixed since [8u212](JDK-8222858) but we had better have this as @HeartSaVioR commented before. Since both classes are `private`, I think the change will be okay. Thank you, @shanyu and @HeartSaVioR !

cc @vanzin

File: core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala [hunk 5]
```diff
@@ -114,6 +114,11 @@ private[history] abstract class ApplicationHistoryProvider {
    */
   def stop(): Unit = { }
 
+  /**
+   * Called when the server is starting up.
```
[the remark this answers is not in the document]
Comment (shanyu) [reply core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala]: @dongjoon-hyun This is the same as stop() that is called by the server (HistoryServer), start() should be called during server starting up. I'll add a few more sentence to describe that it should do and why we added it.

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/history/ApplicationHistoryProvider.scala]: Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM again. Thank you, @shanyu and @HeartSaVioR .
Since it's difficult to add a deadlock test case here, this look okay to me.
Merged to master.

**Pull request closed without merging**